twilight-gateway/src/error.rs

```diff
@@ -0,0 +1,481 @@
+    pub(crate) fn from_compression(source: CompressionError) -> Self {
+        Self {
+            kind: ProcessErrorType::Compression,
+            source: Some(Box::new(source)),
+        }
+    }
+
+    /// Shortcut to create a new error from a gateway event parsing error.
+    pub(crate) fn from_json(source: GatewayEventParsingError) -> Self {
+        Self {
+            kind: ProcessErrorType::Deserializing,
+            source: Some(Box::new(source)),
+        }
+    }
+
+    /// Shortcut to create a new error from a message sending error.
+    pub(crate) fn from_send(source: SendError) -> Self {
+        Self {
+            kind: ProcessErrorType::SendingMessage,
+            source: Some(Box::new(source)),
+        }
+    }
+}
+
+impl Display for ProcessError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self.kind {
+            ProcessErrorType::Compression => {
+                f.write_str("compression failed because the payload may be invalid")
+            }
+            ProcessErrorType::Deserializing => {
+                f.write_str("payload isn't a recognized gateway event")
+            }
+            ProcessErrorType::ParsingPayload => f.write_str("payload could not be parsed as json"),
+            ProcessErrorType::SendingMessage => {
+                f.write_str("failed to send a message over the websocket")
+            }
+        }
+    }
+}
+
+impl Error for ProcessError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        self.source
+            .as_ref()
+            .map(|source| &**source as &(dyn Error + 'static))
+    }
+}
+
+/// Type of [`ProcessError`] that occurred.
+#[derive(Debug)]
+pub enum ProcessErrorType {
+    /// Message could not be decompressed.
+    Compression,
+    /// Received gateway event failed to be deserialized.
+    ///
+    /// The message payload is likely an unrecognized type that is not yet
+    /// supported.
+    Deserializing,
+    /// There was an error parsing a GatewayEvent payload.
+    ParsingPayload,
+    /// Message could not be sent over the Websocket connection.
+    ///
+    /// This may happen when the shard sends heartbeats or attempts to identify
+    /// a new gateway session.
+    SendingMessage,
+}
+
+/// Receiving the next Websocket message failed.
+#[derive(Debug)]
+pub struct ReceiveMessageError {
+    /// Type of error.
+    pub(crate) kind: ReceiveMessageErrorType,
+    /// Source error if available.
+    pub(crate) source: Option<Box<dyn Error + Send + Sync>>,
+}
+
+impl ReceiveMessageError {
+    /// Whether the error is fatal.
+    ///
+    /// If the error is fatal then further attempts to use the shard will return
+    /// more fatal errors.
+    pub fn is_fatal(&self) -> bool {
+        if let ReceiveMessageErrorType::FatallyClosed { close_code } = self.kind() {
+            !CloseCode::try_from(*close_code).map_or(false, CloseCode::can_reconnect)
```

Comment:
Don't know what the intent is, but `!` should only invert the happy case
```suggestion
            CloseCode::try_from(*close_code).map_or(false, !CloseCode::can_reconnect)
```